src/canopy/llm/openai.py

```diff
@@ -206,3 +207,135 @@ async def agenerate_queries(self,
+            model_params: Model parameters to use for this request. Defaults to None (uses the default model parameters).
+                          Overrides the default model parameters if set on initialization.
+                          For example, you can pass: {"temperature": 0.9, "top_p": 1.0} to override the default temperature and top_p.
+                          see: https://platform.openai.com/docs/api-reference/chat/create
+
+        Returns:
+            dict: Function call arguments as a dictionary.
+
+        Usage:
+            >>> from canopy.llm import AzureOpenAILLM
+            >>> from canopy.llm.models import Function, FunctionParameters, FunctionArrayProperty
+            >>> from canopy.models.data_models import UserMessage
+            >>> llm = AzureOpenAILLM()
+            >>> messages = [UserMessage(content="I was wondering what is the capital of France?")]
+            >>> function = Function(
+            ...     name="query_knowledgebase",
+            ...     description="Query search engine for relevant information",
+            ...     parameters=FunctionParameters(
+            ...         required_properties=[
+            ...             FunctionArrayProperty(
+            ...                 name="queries",
+            ...                 items_type="string",
+            ...                 description='List of queries to send to the search engine.',
+            ...             ),
+            ...         ]
+            ...     )
+            ... )
+            >>> llm.enforced_function_call(messages, function)
+            {'queries': ['capital of France']}
+        """  # noqa: E501
+
+        model_params_dict: Dict[str, Any] = deepcopy(self.default_model_params)
+        model_params_dict.update(model_params or {})
+
+        function_dict = cast(ChatCompletionToolParam, function.dict())
+
+        chat_completion = self._client.chat.completions.create(
+            model=self.model_name,
+            messages=[m.dict() for m in messages],
+            functions=[function_dict],
+            function_call={"name": function.name},
+            max_tokens=max_tokens,
+            **model_params_dict,
+        )
+
+        result = chat_completion.choices[0].message.function_call.arguments
+        arguments = json.loads(result)
+
+        jsonschema.validate(instance=arguments, schema=function.parameters.dict())
+        return arguments
+
+    async def achat_completion(
+        self,
+        messages: Messages,
+        *,
+        stream: bool = False,
+        max_generated_tokens: Optional[int] = None,
+        model_params: Optional[dict] = None,
+    ) -> Union[ChatResponse, Iterable[StreamingChatChunk]]:
+        raise NotImplementedError()
+
+    async def agenerate_queries(
+        self,
+        messages: Messages,
+        *,
+        max_generated_tokens: Optional[int] = None,
+        model_params: Optional[dict] = None,
+    ) -> List[Query]:
+        raise NotImplementedError()
```

Comment:
Note that the `AzureOpenAILLM` class requires its own system test (since we don't have stand alone unit test for `LLM` classes yet).  
The test can assume that an Azure OpenAI endpoint was created, and that both the endpoint and the API key are provided by env vars.  
Once the test file exists (and passes locally on your branch) - we will add `AZURE_OPENAI_API_KEY` and `AZURE_OPENAI_ENDPOINT` secretes to the repo's CI configuration.